analysis_driver/quality_control/calculate_relatedness.py

```diff
@@ -46,4 +53,138 @@ def run_vcftools(self):
+        return family_id
+
+    def relationship(self, member):
+        relationship = clarity.get_sample(member).udf.get('Relationship')
+        if not relationship:
+            return 'Other'
+        return relationship
+
+    def sex(self, member):
+        sex = clarity.get_sample(member).udf.get('Sex')
+        if not sex:
+            return 'No_Sex'
+        return sex
+
+    def relationships(self, family):
+        relationship_codes = {'Proband': {'Mother':'0', 'Father':'0'},
+                              'Mother':{'Mother':'0', 'Father':'0'},
+                              'Father':{'Mother':'0', 'Father':'0'},
+                              'Sister':{'Mother':'0', 'Father':'0'},
+                              'Brother': {'Mother':'0', 'Father':'0'},
+                              'Other':{'Mother':'0', 'Father':'0'}}
+
+        for member in family:
+            relationship = self.relationship(member)
+            if relationship in ['Father', 'Mother']:
+                    for i in ['Proband', 'Sister', 'Brother']:
+                        relationship_codes[i][relationship] = member
+        return relationship_codes
+
+    def get_member_details(self, family, all_families):
+        family_lines = []
+        family_info = self.relationships(all_families[family])
+        for member in all_families[family]:
+            sex_codes = {'Male': '1', 'Female': '2', 'No_Sex': '0'}
+            relationship = self.relationship(member)
+            member_id = clarity.get_user_sample_name(member)
+            if not family_info[relationship]['Father'] == '0':
+                father_id = clarity.get_user_sample_name(family_info[relationship]['Father'])
+            else:
+                father_id = family_info[relationship]['Father']
+            if not family_info[relationship]['Mother'] == '0':
+                mother_id = clarity.get_user_sample_name(family_info[relationship]['Mother'])
+            else:
+                mother_id = family_info[relationship]['Mother']
+
+            line = [family,
+                    member_id,
+                   father_id,
+                   mother_id,
+                   sex_codes[self.sex(member)],
+                   '0']
+            family_lines.append(line)
+        return family_lines
+
+    def all_families(self):
+        all_families = {}
+        seen_user_samples = []
+        for i in self.ids:
+            if not clarity.get_user_sample_name(i) in seen_user_samples:
+                seen_user_samples.append(clarity.get_user_sample_name(i))
+                family_id = self.family_id(i)
+                if not all_families.get(family_id):
+                    all_families[family_id] = []
+                all_families[family_id].append(i)
+        return all_families
+
+    @property
+    def ped_file_content(self):
+        ped_file_content = []
+        for family in self.all_families():
+            family_lines = self.get_member_details(family, self.all_families())
```

Comment:
You're calling `self.all_famillies()` twice which means the function will run twice.
You could store it in `ped_file_content` or cache it in `all_families` making it a `@property`